src/nvim/search.c

```diff
@@ -653,7 +653,7 @@ int searchit(
               }
               if (matchcol == 0 && (options & SEARCH_START))
                 break;
-              if (ptr[matchcol] == NUL
+              if (STRLEN(ptr) <= (size_t)matchcol || ptr[matchcol] == NUL
```

Comment:
Did you see the comment below(at `ml_get_buf`)? Even though it is theoretically getting the same line, is it guaranteed that `ptr` will point to the same string? I'd rather not optimize this prematurely and risk introducing bugs, especially because vim is too complex and hard to have its state predicted after a non-trivial function is called(`vim_regexec_multi`)